crates/ruff_python_ast/src/visitor/preorder.rs

```diff
@@ -0,0 +1,1101 @@
+use crate::prelude::*;
+
+/// Visitor that traverses all nodes recursively in pre-order.
+pub trait PreorderVisitor<'a> {
```

Comment:
If there are benefits to doing the preorder traversal, I feel like we should move more of our other visitors to use it. E.g., `StatementVisitor` could probably be preorder... as could most of the implementations of `Visitor` (apart from the AST checker).

Reply:
Maybe I should remove that recommendation again. The pre-order visitation can be slower because of the AST's design (requires zipping in some situations). But then, the out-of-order visitation may be unexpected for users (I was surprised by it). 

I don't plan to refactor any existing code because I want to focus on the formatter. We can follow up on this if it comes up more often.

Reply:
All good, trust you